### PR TITLE
Module: ML Infra Projects

### PR DESCRIPTION
**Pull request by caetano-colin**

### CODE REVIEW

File: 4-projects/modules/ml_single_project/remote.tf [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  environment_kms_project_id = try(data.terraform_remote_state.environments_env.outputs.env_kms_project_id, "")
```
Comment (caetano-colin) [inline 4-projects/modules/ml_single_project/remote.tf]: virar parâmetro

**Comment by daniel-cit** [comment]
Check the output for diffs and correct using terraform fmt <dir>

File: 4-projects/modules/ml_single_project/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/README.md:6]: Please add a `#` with the full name of module and a paragraph with a brief explanation

File: 4-projects/modules/ml_single_project/main.tf [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+module "project" {
+  source  = "terraform-google-modules/project-factory/google"
+  version = "~> 14.1"
+
+  random_project_id        = true
+  random_project_id_length = 4
+  activate_apis            = distinct(concat(var.activate_apis, ["billingbudgets.googleapis.com"]))
+  name                     = "${var.project_prefix}-${local.env_code}-${var.business_code}${var.project_suffix}"
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/main.tf]: this could be a variable

File: 4-projects/modules/ml_single_project/main.tf [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+// Add key for project
+resource "google_kms_crypto_key" "kms_keys" {
+  for_each        = toset(var.key_rings)
+  name            = module.project.project_name
+  key_ring        = each.key
+  rotation_period = var.key_rotation_period
+  lifecycle {
+    prevent_destroy = false
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/main.tf]: this could be a variable

File: 4-projects/modules/ml_single_project/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "project_id" {
+  description = "Project sample project id."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/outputs.tf]:
```suggestion
  description = "Project ID."
```

File: 4-projects/modules/ml_single_project/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  value       = module.project.project_id
+}
+
+output "sa" {
+  description = "Project SA email"
+  value       = module.project.service_account_email
+}
+
+output "project_number" {
+  description = "Project sample project number."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/outputs.tf]:
```suggestion
  description = "Project number."
```

File: 4-projects/modules/ml_single_project/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  value       = module.project.project_number
+}
+
+output "enabled_apis" {
+  description = "VPC Service Control services."
+  value       = distinct(concat(var.activate_apis, ["billingbudgets.googleapis.com"]))
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/outputs.tf]: this value should be in a local to be used in both places this is used

File: 4-projects/modules/ml_single_project/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+output "project_name" {
+  description = "Name of the Project."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/outputs.tf]:
```suggestion
  description = "Project Name."
```

File: 4-projects/modules/ml_single_project/variables.tf [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "org_id" {
+  description = "The organization id for the associated services"
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/variables.tf]:
```suggestion
  description = "The Organization ID"
```

File: 4-projects/modules/ml_single_project/variables.tf [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+variable "activate_apis" {
+  description = "The api to activate for the GCP project"
+  type        = list(string)
+  default     = []
+}
+
+variable "environment" {
+  description = "The environment the single project belongs to"
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/variables.tf]:
```suggestion
  description = "The environment the project belongs to"
```

File: 4-projects/modules/ml_single_project/variables.tf [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  type        = string
+}
+
+variable "vpc_type" {
+  description = "The type of VPC to attach the project to. Possible options are base or restricted."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/variables.tf]:
```suggestion
  description = "The type of VPC to attach the project to. Possible options are `base` or `restricted`."
```

Comment (daniel-cit) [reply 4-projects/modules/ml_single_project/variables.tf]: description needs to say what happens if the value is `""`

File: 4-projects/modules/ml_single_project/variables.tf [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+  type        = list(string)
+}
+
+variable "remote_state_bucket" {
+  description = "Backend bucket to load Terraform Remote State Data from previous steps."
+  type        = string
+}
+
+variable "default_service_account" {
+  description = "Project default service account setting: can be one of `delete`, `depriviledge`, or `keep`."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_single_project/variables.tf]: description need to be updated with the current possible values. the default value is not one of the options listed

File: 4-projects/modules/ml_infra_projects/outputs.tf [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "service_catalog_project_id" {
+  description = "Service Catalog Project ID."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_infra_projects/outputs.tf:18]: for this description and all descriptions, both in variables and outputs, a decision must be made if the description will end with a `.` or not.

**Comment by caetano-colin** [comment]
> Check the output for diffs and correct using terraform fmt

What do you mean by "diffs"?

About terraform fmt:

I opened a PR to fix the build error:
https://github.com/GoogleCloudPlatform/terraform-google-enterprise-genai/pull/39

When merging these changes to this branch, build is green:
https://github.com/caetano-colin/terraform-google-enterprise-genai/pull/5/commits/be0ba50b372727917e4f98586d169353622a1cae

I reverted the change above on this PR to keep the scope closed, as you can see in commit 
https://github.com/caetano-colin/terraform-google-enterprise-genai/pull/5/commits/38069e26fd9fc380469dd86301fb935b3bb9555f

File: 4-projects/modules/ml_infra_projects/variables.tf [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "org_id" {
+  description = "The organization id for the associated services"
```
Comment (daniel-cit) [inline 4-projects/modules/ml_infra_projects/variables.tf]:
```suggestion
  description = "The Organization ID"
```

File: 4-projects/modules/ml_infra_projects/variables.tf [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+variable "secondary_contact" {
+  description = "The secondary email contact for the project"
+  type        = string
+  default     = ""
+}
+
+variable "vpc_type" {
+  description = "The type of VPC to attach the project to. Possible options are base or restricted."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_infra_projects/variables.tf]:
```suggestion
  description = "The type of VPC to attach the project to. Possible options are `base` or `restricted`."
```
description needs to say what happens if the value is ""

File: 4-projects/modules/ml_infra_projects/variables.tf [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+variable "key_rotation_period" {
+  description = "Rotation period in seconds to be used for KMS Key"
+  type        = string
+  default     = "7776000s"
+}
+
+variable "default_service_account" {
+  description = "Project default service account setting: can be one of `delete`, `depriviledge`, or `keep`."
```
Comment (daniel-cit) [inline 4-projects/modules/ml_infra_projects/variables.tf]: description need to be updated with the current possible values. the default value is not one of the options listed

**Review by daniel-cit — CHANGES_REQUESTED** [review]
some minor fixes

File: 4-projects/modules/ml_infra_projects/variables.tf [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+
+variable "remote_state_bucket" {
+  description = "Backend bucket to load Terraform Remote State Data from previous steps."
+  type        = string
+}
+
+variable "business_code" {
+  description = "The code that describes which business unit owns the project"
+  type        = string
+  default     = "abcd"
```
Comment (daniel-cit) [inline 4-projects/modules/ml_infra_projects/variables.tf]: This is not the default value in the foundation
https://github.com/terraform-google-modules/terraform-example-foundation/blob/master/4-projects/modules/single_project/variables.tf#L61

**Pull request closed without merging**